extension/src/lib/portBinding.js

```diff
@@ -13,51 +13,304 @@
 /* global fluid, chrome */
 "use strict";
 
-(function ($, fluid) {
+(function (fluid) {
     var gpii = fluid.registerNamespace("gpii");
 
-    // The main component to handle settings that require DOM manipulations.
-    // It contains various subcomponents for handling various settings.
+    /*
+        The `gpii.chrome.portBinding` grade provides a system for creating a port connection and sending/receiving
+        messages across that port. To verify that messages sent across a port are received and acted upon correctly,
+        posts provide a promise that is resolved/rejected based on a returned receipt. Connections are also set to
+        return receipts after an incoming message has been handled.
+
+        Workflow:
+        post - ( sends payload with message id )
+             - message id gets stored in a map with the promise to resolve/reject
+             -  receiver takes action of the message and returns a receipt ( includes the same message id )
+
+        receive - (receives a message)
+                - if it is a returned receipt
+                  - look into message map and resolve/reject corresponding promise
+                  - remove from map
+                - if it is a new message
+                  - take action on the message
+                  - send a return message (receipt) containing the same message id
+
+
+        Post Messages:
+        1) read/write message
+           - wait for receipt
+        2) read-receipt/write-receipt
+           - do not wait for receipt
+
+        Received Messages:
+        1) read/write message
+           - action, return receipt
+        2) read-receipt/write-receipt
+           - remove from message map
+
+        By default the following message types are handled.
+        - "UIO_PLUS_READ_REQUEST"
+        - "UIO_PLUS_READ_RECEIPT"
+        - "UIO_PLUS_WRITE_REQUEST"
+        - "UIO_PLUS_WRITE_RECEIPT"
+    */
+
     fluid.defaults("gpii.chrome.portBinding", {
         gradeNames: ["fluid.component"],
-        connectionName: "",
+        // Name of the port connection. Will be sent as the `name` when a port connection is created.
+        portName: "",
         members: {
-            lastIncomingMessage: {}
+            port: {
+                expander: {
+                    func: "{that}.setPort"
+                }
+            },
+            openRequests: {}
         },
         events: {
-            onIncomingMessage: null
+            onIncoming: null,
+            onIncomingRead: null,
+            onIncomingReadReceipt: null,
+            onIncomingWrite: null,
+            onIncomingWriteReceipt: null,
+            onDisconnect: null
+        },
+        // defines which event handles which a message type
+        // message types that aren't defined here are ignored.
+        messageForwardingMap: {
+            "UIO_PLUS_READ_REQUEST": "onIncomingRead",
```

Comment:
The capitals and underscores convention looks odd compared to others that we use - and also it differs oddly from the namespacing of the component itself. Perhaps something like gpii.chrome.readRequest for the names?
I know that we use capital letters in the ChangeApplier messages but this is a decision that has been widely criticised and continues only through inertia...

Reply:
@amb26 won't the "." in the name cause issues for the fluid.get call that is used to lookup the mapping?

Reply:
I don't think there's a need for a fluid.get call since the mapping must always be there